[#8707] fix for downgrade of alembic revision 4a5e3465beb6

JamieBeverley opened the pull request:
Currently, downgrading from db revision 4a5e3465beb6 fails because alembic struggles to remove an unnamed foreign key.

This change ensures this foreign key relation is named so that it can be removed using the name, and fixes the downgrade migration by inspecting the fk constraint name.

Fixes #8707 

### Proposed fixes:

Approach taken here:
- added a name to the `ForeignKey(name=...)` and then re-generated the migration with
```
ckan -c /etc/ckan/default/ckan.ini generate migration --autogenerate
```
- manually kept migration revision ids consistent (for folks who've upgraded already)
- altered the `op.drop_constraint` call to pass the name of the fk constraint which is inspected in a helper function (this helper func was added so if someone has upgraded already, they should still be able to downgrade w/ this patch)

(small fix to docs re: generating migrations)

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [x] includes bugfix for possible backport

Please [X] all the boxes above that apply

JamieBeverley commented on ckan/migration/versions/105_4a5e3465beb6_autogenerate_sync.py:
```diff
@@ -56,7 +57,7 @@ def downgrade():
                   postgresql.TIMESTAMP(), autoincrement=False,
                   nullable=True))
-    op.drop_constraint(None, 'resource', type_='foreignkey')
+    op.drop_constraint('resource_package_id_fkey','resource', type_='foreignkey')
```
On second thought... it may just be cleaner to inspect this constraint name (for those who already migrated before this bugfix). This might work instead:
```python
from ckan import model
from sqlalchemy.engine.reflection import Inspector

def get_foreign_key_constraint_name(op):
    table_name = model.resource_table.name
    column_name = next((col.name for col in model.resource_table.columns if col.foreign_keys), None)
    referred_table = model.package_table.name
    
    if not column_name:
        raise ValueError(f"No foreign key column found in {table_name}")

    bind = op.get_bind()
    inspector = Inspector.from_engine(bind)
    fkeys = inspector.get_foreign_keys(table_name)

    for fk in fkeys:
        if fk["referred_table"] == referred_table and fk["constrained_columns"] == [column_name]:
            return fk["name"]
    
    raise ValueError(f"No foreign key constraint found for {table_name}.{column_name} → {referred_table}")


def downgrade():
    # ...
    op.drop_constraint(get_foreign_key_constraint(op), 'resource', type_='foreignkey')
```

JamieBeverley replied:
resolving as PR switched to use this approach

wardi commented:
Thanks @JamieBeverley,

The problem with updating an existing migration is that sites that have already run the migration will have a database that doesn't match the migrations after upgrading.

Let's fix the downgrade script by using plain SQL to remove the foreign key instead.

If you would like to *also* create a new migration that replaces the unnamed foreign key with a named one that could be done here or in a separate PR.

JamieBeverley commented:
Thanks @wardi - I see what you mean, cleaner to keep the forward migration consistent so different folks don't have different db states.

I think I can get to these changes later today!

JamieBeverley commented:
@wardi I switched the patch in `downgrade()` to use plain SQL and removed the naming of the fk constraint for now.

Happy to open another PR with that change if you think its worthwhile but an immediate use-case/need doesn't come to mind.

Thanks again for the feedback
